pydantic/_internal/_generate_schema.py

```diff
@@ -1741,8 +1741,21 @@ def _pattern_schema(self, pattern_type: Any) -> core_schema.CoreSchema:
             raise PydanticSchemaGenerationError(f'Unable to generate pydantic-core schema for {pattern_type!r}.')
 
     def _hashable_schema(self) -> core_schema.CoreSchema:
+        # json cannot be validated by isinstance, so we manually construct a schema
+        json_schema = core_schema.union_schema(
+            [
+                core_schema.str_schema(),
+                core_schema.int_schema(),
+                core_schema.float_schema(),
+                core_schema.bool_schema(),
+                core_schema.none_schema(),
+            ]
+        )
         return core_schema.custom_error_schema(
-            core_schema.is_instance_schema(collections.abc.Hashable),
+            core_schema.json_or_python_schema(
+                json_schema=json_schema,
+                python_schema=core_schema.is_instance_schema(collections.abc.Hashable),
+            ),
```

Comment:
We could use a custom `serialization` schema, see https://github.com/pydantic/pydantic-core/blob/57e69918670016c4cf00a844aabcf76bb0cd458f/python/pydantic_core/core_schema.py#L427

Reply:
Here's an example: https://github.com/pydantic/pydantic/blob/1e413857f48a807d3c81281b52c571f07c7077ee/pydantic/_internal/_generate_schema.py#L1703-L1708